controller/controlplane/controller.go

```diff
@@ -211,11 +211,17 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	crbFinalizerSet := controllerutil.AddFinalizer(cp, string(ControlPlaneFinalizerCleanupClusterRoleBinding))
 	vwcFinalizerSet := controllerutil.AddFinalizer(cp, string(ControlPlaneFinalizerCleanupValidatingWebhookConfiguration))
 	if crFinalizerSet || crbFinalizerSet || vwcFinalizerSet {
-		log.Trace(logger, "Setting finalizers", cp)
+		log.Trace(logger, "setting finalizers", cp)
 		if err := r.Client.Update(ctx, cp); err != nil {
+			if k8serrors.IsConflict(err) {
+				log.Debug(logger, "conflict found when updating ControlPlane, retrying", cp)
+				return ctrl.Result{Requeue: true, RequeueAfter: requeueWithoutBackoff}, nil
+			}
 			return ctrl.Result{}, fmt.Errorf("failed updating ControlPlane's finalizers : %w", err)
 		}
-		return ctrl.Result{}, nil
+		// Requeue to ensure that we do not miss next reconciliation request in case
+		// AddFinalizer calls returned true but the update resulted in a noop.
```

Comment:
Can you elaborate on when this could happen? Is it that when finalizers are added in parallel, before a call to `Client.Update`, `Client.Update` will return no error then?

Reply:
We should in theory be guarded against that by https://github.com/kubernetes-sigs/controller-runtime/blob/cb5fcace24fe2b6bcfe94f6b7fe143bd7e8ea2ef/pkg/config/controller.go#L35-L36 which by default is 1 and we [do not change it](https://github.com/Kong/gateway-operator/blob/30d829f9537f9b47d0ca495b086081ea05f76ce0/modules/manager/run.go#L158-L173) but there might also be other, external factors which might cause a change. Hypothetically.

Reply:
Makes sense, thanks 👍